### PR TITLE
Use sparse neighbor communication to send minimal necessary qp data and add threading

### LINKED ISSUE #314: Improve convolution system scalability
Communicate just what is necessary rather than _all_ quadrature point data elements. This could be done by identifying all processor boundary nodes for a given neighboring domain, using them as centers of a kd-tree radius search, and communication the set union of the search results to said neighboring process.

### LINKED ISSUE #320: Make RadialGreensConvolution threaded
Unfortunately this is necessary, as the domain decomposition is limited by the support of the greens function (i.e. the cutoff radius).

### PR DESCRIPTION
**Pull request by dschwen**
Instead of broadcasting the entire sample's qp data set to every rank this change cause only the necessary amount qp data to be transferred between neighboring processors.

The algorithm fetches a list of point neighbors of the current processor's element set (only done on mesh changes), followed by kd-tree searches with radius `r_cut` plus half the elements diameter to assemble a set of QPs to send to the adjacent processor (done either on mesh changes or at every invocation if a displaced mesh is detected).

Finding pointneighbors of the current processor's element set involves building a node to element map and a periodic node map. The latter (provided by MooseMesh) is unfortunately incomplete, as it does not correctly treat corner and edge nodes that are subject to multiple PBCs. This is addressed in `findNotLocalPeriodicPointNeighbors`. For periodic point neighbors we also store a periodic translation vector to get the correct kd-tree search center (hence the two members `_point_neighbors ` and `_periodic_point_neighbors `).

Closes #314

Also make the convolution loop threaded

Closes #320

### CODE REVIEW

**Comment by dschwen** [comment]
I might add that the algorithm is not perfect yet, but works well for sufficiently compact partittionings. Take a partitioning like the one below

```
_____________________
|       | |          |
|       | |          |
|    1  |2|     3    |
|       | |          |
|       | |          |
_____________________
```

The elements on the left side of rank 3 are not point neighbors of rank 1. In this scenario only ran2 will be sent data from the right edge of rank 1. If `r_cut` is larger than the thickness of 2 rank 3 will be missing data required for the convolution.

This could be remedied by providing a geometric relationship manager that makes sure all elements that touch a processor boundary are ghosted.

**Comment by dschwen** [comment]
Argh, this requires idaholab/moose#11896, but that still fails on intel

File: include/other/ThreadedRadialGreensConvolutionLoop.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  /// name of the Green's Function
+  FunctionName _function_name;
+
+  /// integral over the convolution contribution
+  Real _convolution_integral;
+
+  /// ID number of the current thread
+  THREAD_ID _tid;
+
+private:
```
Comment (snschune) [inline include/other/ThreadedRadialGreensConvolutionLoop.h]: remove?

File: src/userobjects/RadialGreensConvolution.C [hunk 2]
```diff
@@ -406,3 +360,242 @@ RadialGreensConvolution::threadJoin(const UserObject & y)
...
+          // add all face node touching elements directly
+          for (unsigned int n = 0; n < (*it)->n_nodes(); ++n)
+            if ((*it)->is_node_on_side(n, s))
+              insertNotLocalPointNeighbors((*it)->node_id(n));
+        }
+      }
+      else
+      {
+        // find periodic node neighbors and
+        for (unsigned int n = 0; n < (*it)->n_nodes(); ++n)
```
[the remark this answers is not in the document]
Comment (dschwen) [reply src/userobjects/RadialGreensConvolution.C]: Yes, `periodic == nullptr` below

File: src/userobjects/RadialGreensConvolution.C [hunk 3]
```diff
@@ -182,206 +187,155 @@ RadialGreensConvolution::execute()
...
+
+    // data structures for sparse point to point communication
+    std::vector<std::vector<QPData>> send(non_zero_comm.size());
+    std::vector<Parallel::Request> send_requests(non_zero_comm.size());
+    Parallel::MessageTag send_tag = _communicator.get_unique_tag(4711);
+    std::vector<QPData> receive;
+
+    const auto item_type = StandardType<QPData>(&(_qp_data[0]));
+
+#if 0
```
[the remark this answers is not in the document]
Comment (dschwen) [reply src/userobjects/RadialGreensConvolution.C]: no, I still need this to create a few graphs for the paper, then I'll remove it

File: src/userobjects/RadialGreensConvolution.C [hunk 4]
```diff
@@ -182,206 +187,155 @@ RadialGreensConvolution::execute()
...
-    // Loop over all data structures for all processors to perform the gather operation
-    std::istringstream iss;
-    for (unsigned int rank = 0; rank < recv_buffers.size(); ++rank)
+    // !!!!!!!!!!!
+    // !!CAREFUL!! Is it guaranteed that _qp_data is in the same order if the mesh has not changed?
+    // According to @friedmud it is not guaranteed if threads are used
+    // !!!!!!!!!!!
+
+    // update after mesh changes and/or if a displaced problem exists
+    if (_update_communication_lists || _fe_problem.getDisplacedProblem() ||
```
[the remark this answers is not in the document]
Comment (dschwen) [reply src/userobjects/RadialGreensConvolution.C]: not super expensive

File: src/userobjects/RadialGreensConvolution.C [hunk 5]
```diff
@@ -7,42 +7,27 @@
...
-  dataLoad(stream, qpd._volume, context);
-  dataLoad(stream, qpd._value, context);
-}
-
 // specialization for PointListAdaptor<RadialGreensConvolution::QPData>
 template <>
-inline const Point &
-PointListAdaptor<RadialGreensConvolution::QPData>::getPoint(const size_t idx) const
+const Point &
+PointListAdaptor<RadialGreensConvolution::QPData>::getPoint(
```
[the remark this answers is not in the document]
Comment (dschwen) [reply src/userobjects/RadialGreensConvolution.C]: Allows to use vectors of arbitrary compound types to be indexed by the k-d tree, as long as they have a point member that specifies the location to be indexed.

File: src/userobjects/RadialGreensConvolution.C [hunk 6]
```diff
@@ -52,14 +37,18 @@ validParams<RadialGreensConvolution>()
...
+  params.addRequiredParam<FunctionName>(
+      "function",
+      "Green's function (distance is substituted for x) without geometrical attenuation");
   params.addRequiredParam<Real>("r_cut", "Cut-off radius for the Green's function");
   params.addParam<bool>("normalize", false, "Normalize the Green's function integral to one");
 
   // we run this object once at the beginning of the timestep by default
   params.set<ExecFlagEnum>("execute_on") = EXEC_TIMESTEP_BEGIN;
 
+  // make sure we have always have geometric point neighbors ghosted
```
[the remark this answers is not in the document]
Comment (dschwen) [reply src/userobjects/RadialGreensConvolution.C]: The relationship manager line below should do that

File: src/userobjects/RadialGreensConvolution.C [hunk 7]
```diff
@@ -406,3 +360,242 @@ RadialGreensConvolution::threadJoin(const UserObject & y)
...
+    kd_tree->radiusSearch(&(centroid(0)), r_cut2 * r_cut2, ret_matches, search_params);
+    for (auto & match : ret_matches)
+      _communication_lists[elem->processor_id()].insert(match.first);
+  }
+
+  // request fulfilled
+  _update_communication_lists = false;
+}
+
+libMesh::Parallel::StandardType<RadialGreensConvolution::QPData>::StandardType(
```
[the remark this answers is not in the document]
Comment (dschwen) [reply src/userobjects/RadialGreensConvolution.C]: lol

File: include/userobjects/RadialGreensConvolution.h [hunk 8]
```diff
@@ -101,13 +120,61 @@ class RadialGreensConvolution : public ElementUserObject
...
+  std::unique_ptr<PointLocatorBase> _point_locator;
+
+  /**
+   * The data structure which is a list of nodes that are constrained to other nodes
+   * based on the imposed periodic boundary conditions.
+   */
+  std::multimap<dof_id_type, dof_id_type> _periodic_node_map;
+
+  /// The data structure used to find neighboring elements give a node ID
+  std::vector<std::vector<const Elem *>> _nodes_to_elem_map;
```
Comment (permcody) [inline include/userobjects/RadialGreensConvolution.h]: MooseMesh::nodeToElemMap?

File: src/other/ThreadedRadialGreensConvolutionLoop.C [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+          if (local_qp._q_point(j) + _green._periodic_vector[j](j) - r_cut <
+              _green._periodic_max[j])
+            new_cell_vector.push_back(cell + _green._periodic_vector[j]);
+
+          if (local_qp._q_point(j) - _green._periodic_vector[j](j) + r_cut >
+              _green._periodic_min[j])
+            new_cell_vector.push_back(cell - _green._periodic_vector[j]);
+        }
+
+        cell_vector.insert(cell_vector.end(), new_cell_vector.begin(), new_cell_vector.end());
```
Comment (permcody) [inline src/other/ThreadedRadialGreensConvolutionLoop.C]: potential optimization: `std::list::splice`

File: src/userobjects/RadialGreensConvolution.C [hunk 10]
```diff
@@ -406,3 +360,239 @@ RadialGreensConvolution::threadJoin(const UserObject & y)
...
   _convolution.insert(uo._convolution.begin(), uo._convolution.end());
 }
+
+void
+RadialGreensConvolution::insertNotLocalPointNeighbors(dof_id_type node)
+{
+  mooseAssert(!_nodes_to_elem_map[node].empty(), "Node not found in _nodes_to_elem_map");
+
+  for (const auto * elem : _nodes_to_elem_map[node])
+    if (elem->processor_id() != _my_pid)
```
Comment (permcody) [inline src/userobjects/RadialGreensConvolution.C]: NVM - looks like your node_to_elem map is different from the one in `MooseMesh`.

File: src/userobjects/RadialGreensConvolution.C [hunk 11]
```diff
@@ -406,3 +360,239 @@ RadialGreensConvolution::threadJoin(const UserObject & y)
...
+
+  // clear point neighbor set
+  _point_neighbors.clear();
+
+  // my processor id
+  const auto my_pid = processor_id();
+
+  // iterate over active local elements
+  const auto end = mesh.active_local_elements_end();
+  for (auto it = mesh.active_local_elements_begin(); it != end; ++it)
```
Comment (permcody) [inline src/userobjects/RadialGreensConvolution.C]: You can use `as_range` if you want to use a range-based for loop.

File: src/userobjects/RadialGreensConvolution.C [hunk 12]
```diff
@@ -406,3 +360,239 @@ RadialGreensConvolution::threadJoin(const UserObject & y)
...
+  StandardType<Point> d1(&example->_q_point);
+  StandardType<dof_id_type> d2(&example->_elem_id);
+  StandardType<short> d3(&example->_qp);
+  StandardType<Real> d4(&example->_volume);
+  StandardType<Real> d5(&example->_value);
+
+  MPI_Datatype types[] = {
+      (data_type)d1, (data_type)d2, (data_type)d3, (data_type)d4, (data_type)d5};
+  int blocklengths[] = {1, 1, 1, 1, 1};
+  MPI_Aint displs[5], start;
```
Comment (permcody) [inline src/userobjects/RadialGreensConvolution.C]: You might consider `std::array` here. It should work even with these MPI calls.

File: src/userobjects/RadialGreensConvolution.C [hunk 13]
```diff
@@ -406,3 +360,239 @@ RadialGreensConvolution::threadJoin(const UserObject & y)
...
+
+  libmesh_call_mpi(MPI_Get_address(const_cast<RadialGreensConvolution::QPData *>(example), &start));
+  libmesh_call_mpi(MPI_Get_address(const_cast<Point *>(&example->_q_point), &displs[0]));
+  libmesh_call_mpi(MPI_Get_address(const_cast<dof_id_type *>(&example->_elem_id), &displs[1]));
+  libmesh_call_mpi(MPI_Get_address(const_cast<short *>(&example->_qp), &displs[2]));
+  libmesh_call_mpi(MPI_Get_address(const_cast<Real *>(&example->_volume), &displs[3]));
+  libmesh_call_mpi(MPI_Get_address(const_cast<Real *>(&example->_value), &displs[4]));
+
+  for (std::size_t i = 0; i < 5; ++i)
```
Comment (permcody) [inline src/userobjects/RadialGreensConvolution.C]: ... Then you wouldn't have to hardcode the size in multiple places.

**Comment by snschune** [comment]
@dschwen  all done?

File: src/other/ThreadedRadialGreensConvolutionLoop.C (hunk 9 above)
Comment (dschwen) [reply src/other/ThreadedRadialGreensConvolutionLoop.C]: Yes! splice turns out to be about twice as fast in this scenario here. I'll make a PR with that change. There are other places in the code where splice could help, too.

**Pull request merged**